jest.config.js

```diff
@@ -10,6 +10,7 @@ const createJestConfig = nextJest({
 });
 const jestConfig = createJestConfig({
   moduleDirectories: ["node_modules", "<rootDir>"],
+  testTimeout: 60000,
```

Comment:
O que acha de transformar isso em alguma conta ou algum comentário para facilizar a leitura do código. Exemplo:

```js
// using coments
testTimeout: 60000,  // 1min

//using calc
testTimeout: 1000 * 60 *  1, 
```

Além isso  esse valor não poderia vir das variaveis de ambiente?

Reply:
A ideia do comentário me parece ok.
Não gosto da ideia de usar a conta, parece que complica mais do que o necessário.
Não sei se faz sentido transformar esse valor em uma variável de ambiente visto que esse valor não tende a ser alterado com tanta frequência.
Eu vivo com os 60000 hehe

Reply:
Eu usei wait-on  [wait-on](https://www.npmjs.com/package/wait-on), no meu caso resolveu.
`"test": "npm run services:up && npm run wait-for-postgres && concurrently -n next,jest -k -s command-jest 'next dev' 'wait-on http://localhost:3000 && jest --runInBand'"`

Reply:
Acho que os pontos mais essenciais já foram levantados, então, vou deixar apenas uma sugestão :+1:

```js
const TIMEOUT_IN_MILLISECONDS = 60_000;

// ...
testTimeout: TIMEOUT_IN_MILLISECONDS,
// ...
```

Reply:
Me incomoda o fato de aumentarmos o `timeout` aqui pelas configurações gerais porque acredito que assim podemos sofrer outros impactos negativos no futuro, como acabar não identificando testes lentos que poderiam ser otimizados ou ocultando problemas de performance, por exemplo.
Pesquisando um pouco, cheguei em uma possível solução e queria validar com vocês se faz sentido 😊 

No caso o `jest` possui uma configuração [setupFilesAfterEnv](https://jestjs.io/docs/configuration#setupfilesafterenv-array) que permite a gente executar códigos logo depois do framework ter sido instalado no ambiente, mas antes dos códigos de testes em si. Pensando na utilização desse carinha, podíamos criar um arquivo de `setup` contendo a lógica do `beforeAll` a ser utilizada por todos, e de quebra especificar o `timeout` de 1min apenas para esse processo de inicialização:

````js
// jest.setup.js

import orchestrator from "tests/orchestrator.js";

beforeAll(async () => {
  await orchestrator.waitForAllServices();
}, 60000);

````


````js
// jest.config.js
const nextJest = require("next/jest");
const dotenv = require("dotenv");

dotenv.config({ path: ".env.development" });
const createJestConfig = nextJest({ dir: "." });

const jestConfig = createJestConfig({
  moduleDirectories: ["node_modules", "<rootDir>"],
  setupFilesAfterEnv: ["<rootDir>/jest.setup.js"],
  // testTimeout: 60000,
});

module.exports = jestConfig;
````

Assim não teríamos a necessidade de chamar o orquestrador em todos os arquivos de teste e ainda não precisaríamos aumentar o `timeout` máximo geral. Na minha cabeça tá fazendo sentido kkkkkk o que acham?

Reply:
@nadineouro segundo a documentação do Jest, os arquivos contidos em [`setupFilesAfterEnv`](https://jestjs.io/docs/29.6/configuration#setupfilesafterenv-array) rodarão antes de cada suíte de testes:

> A list of paths to modules that run some code to configure or set up the testing framework **before each test file** in the suite is executed

![image](https://github.com/filipedeschamps/clone-tabnews/assets/57193392/601ded58-cf17-4dca-af6b-c42fadb7d2bf)

Com isso, não mudaria o comportamento que temos atualmente apenas deixaríamos centralizado o `beforeAll` com o `orchestrator` em um único arquivo, ou seja, a configuração de `timout` ainda seria global.

---

Sobre centralizar o `orchestrator` em um arquivo, poderia penalizar futuros testes (de unidade, por exemplo) que não necessariamente dependeriam dos serviços que o `orchestrator` interage.

Para exemplificar, vou citar o componente `infra/under-maintenance.js` do TabNews:

Os testes dele não dependem de nada do `orchestrator`, veja:

https://github.com/filipedeschamps/tabnews.com.br/blob/4d62fc6d87dcbab578dd26349800d3475b003a88/tests/unit/infra/under-maintenance.test.js#L8-L10

---

Fez sentido turma?

Reply:
@kaique-soares O que quis dizer é que fora essa inicialização, os testes em si vão continuar com o timeout limite de 5s ao invés de 60s... por exemplo nas minhas tentativas aqui eu forcei um delay num teste específico usando essa abordagem que comentei, e como ele passou dos 5s acusou erro de timeout, q era justamente oq eu queria. Pensei nisso pq supondo que a gente acabe desenvolvendo um teste mal estruturado que leva 30s p rodar, ele vai acabar passando despercebido pq o tempo máximo de espera aceito está muito alto. 

Mas sobre a outra questão realmente não faria sentido, quando escrevi estava com a visão limitada apenas aos testes que já tínhamos criado e aí pensei que todos iriam precisar desse inicialização prévia, mas não faz sentido mesmo kkkkk

Reply:
@nadineouro perfeito, muito massa você ter levando esses pontos porque dessa forma aprendemos todos juntos e deixamos registrados os conhecimentos para quem esbarrar com essa PR no futuro :handshake:

Sobre identificar os testes demorados, o Jest tem um retorno visual bem interessante, olha só:

![Screenshot from 2024-06-07 20-31-34](https://github.com/filipedeschamps/clone-tabnews/assets/57193392/6fb4a576-b5a8-4165-9b7b-f31bd6999cf9)

Veja que testes que excederam os `5s` foram destacados com a faixa vermelha, repare o último teste em com a faixa, que preciso :sweat_smile: 

> :warning: No print acima, fiz um checkout para um commit em que o TabNews ainda utilizava Jest porque atualmente utiliza Vitest

A migração ocorreu nessa PR https://github.com/filipedeschamps/tabnews.com.br/pull/1642

Reply:
Após uma investigação encontrei esta solução, e após aplicar ela no meu código os testes passaram.

![image](https://github.com/filipedeschamps/clone-tabnews/assets/102832541/b935c326-8f55-4db8-b181-358aaf1b2583)